src/native/mqtt5_listener.c

```diff
@@ -0,0 +1,599 @@
+        s_aws_count_allocation(event->connack_data->shared_subscriptions_available, &references_needed);
+        s_aws_count_allocation(event->connack_data->subscription_identifiers_available, &references_needed);
+        s_aws_count_allocation(event->connack_data->topic_alias_maximum, &references_needed);
+        s_aws_count_allocation(event->connack_data->wildcard_subscriptions_available, &references_needed);
+        /* Add user properties */
+        references_needed += event->connack_data->user_property_count * 2;
+        references_needed += 2; /* Add 2 for arrays to hold user properties */
+    }
+    if (event->disconnect_data != NULL) {
+        /* A Disconnect packet will need 1 reference at a minimum */
+        references_needed += 1;
+        /* Optionals */
+        s_aws_count_allocation(event->disconnect_data->reason_string, &references_needed);
+        s_aws_count_allocation(event->disconnect_data->server_reference, &references_needed);
+        s_aws_count_allocation(event->disconnect_data->session_expiry_interval_seconds, &references_needed);
+        /* Add user properties */
+        references_needed += event->disconnect_data->user_property_count * 2;
+        references_needed += 2; /* Add 1 for array to hold user properties */
+    }
+    if (event->settings != NULL) {
+        /* Negotiated settings only need 2 references, one for the ClientID and another for the object */
+        references_needed += 2;
+    }
+
+    /* Make a local frame so we can clean memory */
+    jint local_frame_result = (*env)->PushLocalFrame(env, (jint)references_needed);
+    if (local_frame_result != 0) {
+        s_aws_mqtt5_listener_log_and_throw_exception(
+            env, "LifecycleEvent: could not push local JNI frame with 14 allocation minimum!", AWS_ERROR_INVALID_STATE);
+        aws_jni_release_thread_env(jvm, env);
+        return;
+    }
+
+    jobject connack_data = NULL;
+    if (event->connack_data != NULL) {
+        connack_data = s_aws_mqtt5_client_create_jni_connack_packet_from_native(env, event->connack_data);
+        if (connack_data == NULL) {
+            AWS_LOGF_ERROR(AWS_LS_MQTT5_GENERAL, "LifecycleEvent: creating ConnAck packet failed!");
+            goto clean_up;
+        }
+    }
+
+    jobject disconnect_data = NULL;
+    if (event->disconnect_data != NULL) {
+        disconnect_data = s_aws_mqtt5_client_create_jni_disconnect_packet_from_native(env, event->disconnect_data);
+        if (disconnect_data == NULL) {
+            AWS_LOGF_ERROR(AWS_LS_MQTT5_GENERAL, "LifecycleEvent: creating Disconnect packet failed!");
+            goto clean_up;
+        }
+    }
+
+    jobject negotiated_settings_data = NULL;
+    if (event->settings != NULL) {
+        negotiated_settings_data = s_aws_mqtt5_client_create_jni_negotiated_settings_from_native(env, event->settings);
+    }
+
+    jobject jni_lifecycle_events = java_listener->jni_lifecycle_events;
```

Comment:
Extremely minor nitpick: I'd move this check earlier in the function and escape early if the function does not exist. That way there's not a need to do additional allocations and calculations when not necessary.

(I didn't do this for Lifecycle events in MQTT5 either, so... 😓  )